pushsnapscript/snap_store.py

```diff
@@ -2,48 +2,220 @@
+        }
+        for revision, revision_metadata in metadata_per_revision.items()
+        if revision > 1    # First revision doesn't have sha3_384
+    }
+
+
+def _get_from_sha3_384_from_revision(store, revision):
+    # Sadly, this function is not exposed in snapcraft.
+    headers = store.cpi.get_default_headers()
+    headers.update({
+        'Accept': 'application/hal+json',
+        'X-Ubuntu-Series': DEFAULT_SERIES
+    })
+    params = {
+        'fields': 'status,download_sha3_384,revision',
+        'revision': revision
+    }
+    url = 'api/v1/snaps/details/{}'.format(_SNAP_NAME_ON_STORE)
+    resp = store.cpi.get(url, headers=headers, params=params)
+    return resp.json()['download_sha3_384']
+
+
+def _find_revision_and_version_of_current_snap(metadata_per_revision, current_sha3_384):
+    item = get_single_item_from_sequence(
+        metadata_per_revision.items(),
+        lambda item: item[1]['download_sha3_384'] == current_sha3_384,
+        ErrorClass=ValueError,
+        no_item_error_message='No revision has sha3_384 "{}"'.format(current_sha3_384),
+        too_many_item_error_message='Too many revisions have sha3_384 "{}"'.format(current_sha3_384),
+    )
+    revision = item[0]
+    version = item[1]['version']
+    log.debug('Current snap (version "{}") found on the store at revision {}'.format(version, revision))
+    return revision, version
+
+
+def _check_current_snap_is_not_released(current_revision, current_version, latest_released_revision, latest_released_version):
+    if latest_released_revision == current_revision:
+        if latest_released_version == current_version:
+            raise AlreadyLatestError(latest_released_version, latest_released_revision)
+        else:
+            raise TaskVerificationError(
+                'Revisions {} matched but versions differ. Latest on store: "{}". Version of current Snap: "{}"'.format(
+                    latest_released_revision, latest_released_version, current_version
+                )
+            )
+    elif latest_released_version > current_version:
+        # We don't check if the revision is higher because
+        raise TaskVerificationError(
+            'Current version "{}" is lower than the latest one released on the store "{}". Downgrades are not allowed.'.format(
+                current_version, latest_released_version
+            )
+        )
+
+    log.debug(
+        'Current version "{}" is higher than the latest released one "{}". Okay to release the current one'.format(
+            current_version, latest_released_version
+        )
+    )
+
+
+def _pick_revision_and_version_of_latest_released_snap(channel, metadata_per_revision):
+    item = get_single_item_from_sequence(
+        metadata_per_revision.items(),
+        lambda item: channel in item[1]['current_channels'],
+        ErrorClass=ValueError,
+        no_item_error_message='No revision is currently released on channel "{}"'.format(channel),
```

Comment:
Doesn't this mean we can't ever add a new channel without throwing an exception? Is that a concern?

Reply:
That's a very good point! You right, it will bail out. At the moment, I don't foresee the need of creating a new channel. Nightly (even if not shipped yet) will use an existing one. ESR has "esr/stable" and "esr/candidate". 
Worst case scenario, we can still release the new channel manually. It's not too much overhead. I'll keep this concern in mind and revisit if the need comes up.

Reply:
+1, it seems like a rare enough occurrence that we could doc how to set up the channel manually at first.

Reply:
I actually realizes, this needs a manual action anyway. We need to ask Canonical first, like I did in https://forum.snapcraft.io/t/firefox-please-create-the-track-esr/5006